File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> used;

        int l = 0, r = 0, rslt = 0;
        while(r < s.size()){
            if(used[s[r]]){
                while(l < s.size() && s[l] != s[r]) used[s[l++]] = false;
                l++;
                used[s[r]] = false;
            } else{
                used[s[r]] = true;
                rslt = max(rslt, r-l+1);
                r++;
            }
        }

        return rslt;
    }
};

File: Data Structures & Algorithms/house-robber-ii/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:

        def solve(nums: List[int], values: List[int], i, end) -> int:
            if i > end:
                return 0

            if values[i] != -1:
                return values[i]

            values[i] = 0
            plus1 = solve(nums, values, i+1, end)
            plus2 = solve(nums, values, i+2, end) + nums[i]

            values[i] = max(plus1, plus2)
            return values[i]

        start_list = [-1]*len(nums)
        rslt = max(solve(nums, start_list, 0, len(nums)-2), nums[0])
        start_list = [-1]*len(nums)
        rslt = max(rslt, solve(nums, start_list, 1, len(nums)-1))
        
        return rslt
        
            